### DISCUSSION TITLE
Optimizing reading for large collection of images

### DISCUSSION THREAD

**Discussion by wario2k** [dlemstra/Magick.NET#925] [Help]
Hi,
```cs
using (var image = new MagickImage())
                {
                    var readSettings = new MagickReadSettings();
                    image.Read(data, readSettings); // data is a byte array which contains jpeg/jp2 images

                    if (isBgr)
                    {
                        // Convert BGR to RGB
                        var matrix = new MagickColorMatrix(3, 0, 0, 1, 0, 1, 0, 1, 0, 0);
                        image.ColorMatrix(matrix);
                    }

                    image.Format = MagickFormat.Jpg;
                    return image.ToBitmap();
                }
```
I have a large collection of images and I need to decode bytes from a binary file and create bitmaps for manipulation. Is there any way to optimize this process? On average it seems to take around 60-70ms but there are times where it takes over 100ms and considering I have 256+ images to process per iteration it gets really slow. I used a performance profiler and it seems the ToBitmap() and Read() operations are fairly expensive. Are there any faster alternatives to loading images from byte data? I tried running it in parallel loop to help with performance but is still less than ideal. Any ideas would be greatly appreciated. 

Version : Magick.NET-Q8-AnyCPU.7.14.3 
Env: Windows 10
.Net Framework Version: 4.6

Best,
Aayush

**Comment by dlemstra** [comment]
Why do you need `Bitmap`'s for the mutations? This library can do a lot of mutations.

How are you reading the images and why do you need to change `BGR` to `RGB`.

And how often do you need to do this? If it is once I don't think it would be worth the optimization because that would take around 25 seconds?

**Comment by wario2k** [comment]
Hi Dirk,
Thanks for your reply.
1.Bitmaps are needed since this application is translating a lot of legacy
code into C# the application that is consuming the data that I produce
expects Bitmaps so that's primarily the only reason I have it generating
them.
2. I am reading this information from a proprietary binary file which is
just an embedded binary file with a lot of data. Depending on the spec
defined in the file some data is present in RAW, JP2 or JPG formats. There
are various image components that I am using ImageMagick to decode and
store as in-memory objects. The BRG conversion is only applicable to one
specific type of image which was being used as the thumbnail for a
particular report, so that block is not being executed for 95% of the
data.
3. The application has to handle high-volume data conversion for hospitals
so that's why I'm concerned about trying to optimize it as much as
possible. The input file is essentially a binary file that is produced by
some medical device and the tool I am working on reads the binary data
decodes all the information and imports it into our data management system.

**Comment by dlemstra** [comment]
I don't see any options where you could improve. It might be possible that certain formats take more time to read and that is why you are seeing the difference. It might be more interesting to check if you can get rid of the legacy code and translate that to Magick.NET code.
